src/ServiceConnections/RestServiceConnection.cpp

```diff
@@ -0,0 +1,297 @@
+/**
+ * @file RestServiceConnection.cpp
+ * @author Hayden McAfee (hayden@outlook.com)
+ * @version 0.1
+ * @date 2021-02-13
+ *
+ * @copyright Copyright (c) 2020 Hayden McAfee
+ *
+ */
+
+#include "RestServiceConnection.h"
+
+#include "../Utilities/FtlTypes.h"
+
+#include <jansson.h>
+#include <spdlog/spdlog.h>
+#include <sstream>
+
+#pragma region Constructor/Destructor
+RestServiceConnection::RestServiceConnection(
+    std::string hostname,
+    uint16_t port,
+    bool useHttps,
+    std::string authToken)
+:
+    hostname(hostname),
+    port(port),
+    useHttps(useHttps),
+    authToken(authToken)
+{ }
+#pragma endregion
+
+#pragma region Public methods
+void RestServiceConnection::Init()
+{
+    std::stringstream baseUri;
+    baseUri << (useHttps ? "https" : "http") << "://" << hostname << ":" << port;
+    spdlog::info("Using REST Service Connection @ {}", baseUri.str());
+}
+
+Result<std::vector<std::byte>> RestServiceConnection::GetHmacKey(ftl_channel_id_t channelId)
+{
+    std::stringstream url;
+    url << "/hmac/" << channelId;
+
+    httplib::Result response = runGetRequest(url.str());
+    if (response->status >= 400)
+    {
+        return Result<std::vector<std::byte>>::Error(
+            "Channel ID does not have a stream key.");
+    }
+
+    JsonPtr result = decodeRestResponse(response);
+    json_t* jsonStreamKey = json_object_get(result.get(), "hmacKey");
+    if (jsonStreamKey != nullptr && json_is_string(jsonStreamKey))
+    {
+        const char* keyValue = json_string_value(jsonStreamKey);
+        return Result<std::vector<std::byte>>::Success(
+            std::vector<std::byte>(
+                reinterpret_cast<const std::byte*>(keyValue),
+                (reinterpret_cast<const std::byte*>(keyValue) + strlen(keyValue))));
+    }
+
+    return Result<std::vector<std::byte>>::Error(
+        "Could not find a stream key for the given channel.");
+}
+
+Result<ftl_stream_id_t> RestServiceConnection::StartStream(ftl_channel_id_t channelId)
+{
+    std::stringstream url;
+    url << "/start/" << channelId;
+
+    httplib::Result response = runPostRequest(url.str());
+    if (response->status >= 400)
+    {
+        std::stringstream errMsg;
+        errMsg << "Channel " << channelId << " is not allowed to stream.";
+
+        return Result<ftl_stream_id_t>::Error(errMsg.str().c_str());
+    }
+
+    JsonPtr result = decodeRestResponse(response);
+    json_t* jsonStreamId = json_object_get(result.get(), "streamId");
+
+    // TODO: Allow strings and ints from JSON
+    if (jsonStreamId != nullptr && json_is_string(jsonStreamId))
```

Comment:
This was set to accept strings to match the Glimesh Service Connection, but since channel IDs are uints this should really accept JSON numbers (either alongside strings or only numbers). I'm not very familiar with C++ so I've left this for someone else.

Reply:
That's fine - we can open an issue to track this, since the Glimesh service connection should be more safely parsing the channel IDs too.